ntoskrnl/ex/locale.c

```diff
@@ -24,8 +24,252 @@ LANGID PsInstallUILanguageId = LANGIDFROMLCID(0x00000409);
+    RtlInitUnicodeString(&LocaleIdString, ValueIdBuffer);
+    LocaleIdString.MaximumLength = BufferLength;
+    RtlAppendUnicodeToString(&LocaleIdString, LocaleName.Buffer);
+
+    /* Open the NLS locale key */
+    Status = ZwOpenKey(&NlsLocaleKey,
+                       KEY_QUERY_VALUE,
+                       &NlsLocalKeyAttrs);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open %wZ (Status 0x%lx)\n", NlsLocaleKeyPath, Status);
+        return Status;
+    }
+
+    /* Open the NLS alternate sort locales key */
+    Status = ZwOpenKey(&AltSortKey,
+                       KEY_QUERY_VALUE,
+                       &AltSortKeyAttrs);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open %wZ (Status 0x%lx)\n", AltSortKeyPath, Status);
+        goto Quit;
+    }
+
+    /* Open the NLS language groups key */
+    Status = ZwOpenKey(&LangGroupKey,
+                       KEY_QUERY_VALUE,
+                       &LangGroupKeyAttrs);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open %wZ (Status 0x%lx)\n", LangGroupPath, Status);
+        goto Quit;
+    }
+
+    /* Check if the captured locale ID exists in the list of other locales */
+    BufferKey = (PKEY_VALUE_PARTIAL_INFORMATION)ValueBuffer;
+    Status = ZwQueryValueKey(NlsLocaleKey,
+                             &LocaleIdString,
+                             KeyValuePartialInformation,
+                             BufferKey,
+                             sizeof(ValueBuffer),
+                             &ReturnedLength);
+    if (!NT_SUCCESS(Status))
+    {
+        /* We failed, retry by looking at the alternate sorts locales */
+        Status = ZwQueryValueKey(AltSortKey,
+                                 &LocaleIdString,
+                                 KeyValuePartialInformation,
+                                 BufferKey,
+                                 sizeof(ValueBuffer),
+                                 &ReturnedLength);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("Failed to query value from Alternate Sorts key (Status 0x%lx)\n", Status);
+            goto Quit;
+        }
+    }
+
+    /* We got what we need, now query the locale from the language groups */
+    RtlInitUnicodeString(&LocaleIdString, (PWSTR)BufferKey->Data);
+    Status = ZwQueryValueKey(LangGroupKey,
+                             &LocaleIdString,
+                             KeyValuePartialInformation,
+                             BufferKey,
+                             sizeof(ValueBuffer),
+                             &ReturnedLength);
+    if (!NT_SUCCESS(Status) ||
+        BufferKey->DataLength < 2)
+    {
+        DPRINT1("Failed to query value from Language Groups key (Status 0x%lx, DataLength %lu)\n", Status, BufferKey->DataLength);
+        goto Quit;
+    }
+
+    /*
+     * We have queried the locale with data and all that stuff. However we are not
```

Comment:
"with data and all that stuff" --> "with its data." (since it's only what we retrieved, nothing less, nothing more)